### PR TITLE
OAK-9686: replace Elastic RHLC with Elastic Java Client

### PR DESCRIPTION
**Pull request by AngelaFabregues**
Query migration to the new Elasticsearch-Java client library.

### CODE REVIEW

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticTestServer.java [hunk 1]
```diff
@@ -16,18 +16,19 @@
...
 import org.apache.http.impl.client.CloseableHttpClient;
 import org.apache.http.impl.client.HttpClients;
 import org.apache.jackrabbit.oak.commons.IOUtils;
-import org.elasticsearch.Version;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.testcontainers.DockerClientFactory;
 import org.testcontainers.containers.Network;
 import org.testcontainers.elasticsearch.ElasticsearchContainer;
+import org.testcontainers.shaded.org.bouncycastle.asn1.x509.qualified.ETSIQCObjectIdentifiers;
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticTestServer.java]: unused
```suggestion
```

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticTestServer.java [hunk 2]
```diff
@@ -141,4 +142,20 @@ private void checkIfDockerClientAvailable() {
         assumeNotNull(client);
     }
 
+    /**
+     * Launches an Elasticsearch Test Server to re-use among several tests.
+     * @param args
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticTestServer.java]: remove since no args are supported
```suggestion
```

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticTestServer.java [hunk 3]
```diff
@@ -141,4 +142,20 @@ private void checkIfDockerClientAvailable() {
...
+     */
+    public static void main(String[] args){
+        ElasticsearchContainer esContainer = ElasticTestServer.getESTestServer();
+        try {
+            System.out.println("Docker container with Elasticsearch launched at \""+esContainer.getHttpHostAddress()+
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticTestServer.java]:
```suggestion
            System.out.println("Docker container with Elasticsearch launched at \"" + esContainer.getHttpHostAddress() +
```

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticTestServer.java [hunk 4]
```diff
@@ -141,4 +142,20 @@ private void checkIfDockerClientAvailable() {
...
+                "\". Please PRESS ENTER to stop it...");
+            System.in.read();
+            esContainer.stop();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticTestServer.java]: since you are just wrapping the exception in a RuntimeException, you could remove the try-catch block and add the throws in the method declaration

**Review by fabriziofortino — CHANGES_REQUESTED** [review]

File: oak-search-elastic/pom.xml [hunk 5]
```diff
@@ -33,8 +33,8 @@
   <description>Oak Elasticsearch integration subproject</description>
 
   <properties>
-    <elasticsearch.version>7.16.3</elasticsearch.version>
-    <lucene.version>8.10.1</lucene.version>
+    <elasticsearch.version>7.17.3</elasticsearch.version>
+    <jackson.version>2.12.3</jackson.version>
```
Comment (nit0906) [inline oak-search-elastic/pom.xml]: If possible can we keep this in sync with the jackson dependency in the parent pom - https://github.com/apache/jackrabbit-oak/blob/trunk/oak-parent/pom.xml#L66. ? Either downgrade this or upgrade that.

Comment (nit0906) [inline oak-search-elastic/pom.xml]: If possible can we keep this in sync with the jackson dependency in the parent pom - https://github.com/apache/jackrabbit-oak/blob/trunk/oak-parent/pom.xml#L66. ? Either downgrade this or upgrade that.

Comment (nit0906) [reply oak-search-elastic/pom.xml]: ah -  I suppose initially it was picking up the parent pom version - do we specifically need this higher version here ?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexStatistics.java [hunk 6]
```diff
@@ -164,77 +157,59 @@ static <K, V> LoadingCache<K, V> setupCache(long maxSize, long expireMin, long r
...
-
-                if (!indices.isEmpty()) {
-                    // we ask for a specific index, so we can get the first entry
-                    Map<String, String> indexProps = indices.get(0);
-                    String size = indexProps.get("store.size");
-                    String creationDate = indexProps.get("creation.date");
-                    String luceneDocsCount = indexProps.get("docs.count");
-                    String luceneDocsDeleted = indexProps.get("docs.deleted");
-                    return new StatsResponse(
+            IndicesRecord record = crd.connection.getClient().cat().indices(i -> i
```
Comment (nit0906) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexStatistics.java]: Could this be null (similar to the response earlier on ?)
. If yes, then we could possibly have a potential NPE

File: oak-search-elastic/pom.xml (hunk 5 above)
Comment (AngelaFabregues) [reply oak-search-elastic/pom.xml]: Checking it

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexStatistics.java (hunk 6 above)
Comment (AngelaFabregues) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexStatistics.java]: According to the documentation, nulls are prevented by the new library. I will double-check just in case.

Comment (AngelaFabregues) [reply oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexStatistics.java]: You are right, this is a potential null pointer exception. Line 198 ends with getting the first element of a list which can potentially be empty, the index has been deleted.

File: oak-search-elastic/pom.xml (hunk 5 above)
Comment (AngelaFabregues) [reply oak-search-elastic/pom.xml]: I've changed the jackson-databind version to the one in the parent module (using the same variable). I've executed mvn clean install in the module and executed the tests. All tests passed so... I suppose we can use the old version instead of the version provided by the documentation https://www.elastic.co/guide/en/elasticsearch/client/java-api-client/7.17/installation.html#maven
@fabriziofortino , is it ok?

Comment (AngelaFabregues) [reply oak-search-elastic/pom.xml]: @fabriziofortino , do you know why we removed the Lucene version. Or even better, why was it there originally? 
oak-search-elastic tests work well without it.

**Comment by AngelaFabregues** [comment]
@fabriziofortino , I've checked tests while reusing the Elasticsearch instance among test executions. In this case, some tests failed. I skipped two tests from ElasticIndexProviderServiceTest with assumeTrue to avoid null pointer exception. And I left the ElasticReliabilityTest tests even though the assertion fails. Is it enough for this case?

File: oak-search-elastic/pom.xml (hunk 5 above)
Comment (fabriziofortino) [reply oak-search-elastic/pom.xml]: The lucene dependency was required by some methods invoked with the RHLC. These methods have been replaced by the new Java client.

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexProviderServiceTest.java [hunk 7]
```diff
@@ -95,6 +96,8 @@ public void defaultSetup() {
 
     @Test
     public void withElasticSetup() throws Exception {
+        // Skips this test when a single Elasticsearch cluster is re-used among test executionsgit , see the ElasticTestServer main.
+        assumeTrue(elasticRule.useDocker());
```
Comment (fabriziofortino) [inline oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexProviderServiceTest.java]: That's not correct: the tests need to work also with single ES clusters. Please revert these changes. To fix the problem lines 105 and 124 need to be replaced with (I have tried to push the commit myself but I have problem s with this branch):
```java
props.put(PROP_ELASTIC_PORT, elasticRule.getElasticConnectionModel().getElasticPort());
```

**Review by fabriziofortino — CHANGES_REQUESTED** [review]

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexProviderServiceTest.java (hunk 7 above)
Comment (nit0906) [reply oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexProviderServiceTest.java]: The assumeTrue(elasticRule.useDocker()) statement should also be removed from both the tests in which it was added. Otherwise they will only run on a docker based setup.

**Review by nit0906 — CHANGES_REQUESTED** [review]
Minor change requested

File: oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexProviderServiceTest.java (hunk 7 above)
Comment (AngelaFabregues) [reply oak-search-elastic/src/test/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticIndexProviderServiceTest.java]: I removed it

**Review by fabriziofortino — APPROVED** [review]

**Comment by AngelaFabregues** [comment]
> 

I don't know how it got fixed but glad that it is.

**Review by nit0906 — APPROVED** [review]
